crates/rust-analyzer/src/diff.rs

```diff
@@ -0,0 +1,61 @@
+use dissimilar::Chunk;
+use ide::{TextEdit, TextRange, TextSize};
+
+fn textedit_from_chunks(chunks: Vec<dissimilar::Chunk>) -> TextEdit {
+    let mut builder = TextEdit::builder();
+    let mut pos = TextSize::default();
+
+    let mut pending_delete_start = None;
+
+    for chunk in chunks {
+        match (pending_delete_start, chunk) {
+            (Some(delete_start), Chunk::Insert(text)) => {
+                builder.replace(TextRange::new(delete_start, pos), text.into());
+                pending_delete_start = None;
+            }
+            (_, _) => {
+                if let Some(delete_start) = pending_delete_start {
+                    builder.delete(TextRange::new(delete_start, pos));
+                    pending_delete_start = None;
+                }
+
+                match chunk {
+                    Chunk::Equal(text) => {
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Delete(text) => {
+                        pending_delete_start = Some(pos);
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Insert(text) => {
+                        builder.insert(pos, text.into());
+                        pos += TextSize::of(text);
+                    }
+                }
+            }
+        }
+    }
+    if let Some(delete_start) = pending_delete_start {
+        builder.delete(TextRange::new(delete_start, pos));
+    }
+    builder.finish()
+}
+
+pub(crate) fn diff(left: &str, right: &str) -> TextEdit {
```

Comment:
Let's move this fn to top, as per https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#order-of-items